### PR TITLE
Testing updates

### PR DESCRIPTION
**Pull request by courtneyholcomb**
Updates to test data & some tests. These changes are needed up the stack to illustrate how offset metrics work in query output tests.
Definitely recommend reviewing by commit!

### CODE REVIEW

File: tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_to_grain_with_grain_matching_offset__query_output.txt [hunk 1]
```diff
@@ -0,0 +1,260 @@
+test_name: test_offset_to_grain_with_grain_matching_offset
```
Comment (courtneyholcomb) [inline tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_to_grain_with_grain_matching_offset__query_output.txt:1]: not useful but makes sense lol

Comment (courtneyholcomb) [reply tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_to_grain_with_grain_matching_offset__query_output.txt:1]: add a time constraint

File: tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_to_grain_with_custom_grain__query_output.txt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+test_name: test_offset_to_grain_with_custom_grain
```
[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_to_grain_with_custom_grain__query_output.txt:1]: would be more useful with a smaller custom grain

File: tests_metricflow/integration/query_output/test_offset_metrics.py [hunk 3]
```diff
@@ -89,6 +89,7 @@ def test_custom_offset_window_with_base_grain(
     )
 
 
+# TODO: Why is date part allowed here but not on its own?
```
Comment (courtneyholcomb) [inline tests_metricflow/integration/query_output/test_offset_metrics.py:92]: to followup on

File: tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_to_grain_with_grain_smaller_than_offset__query_output.txt [hunk 4]
```diff
@@ -0,0 +1,7678 @@
+test_name: test_offset_to_grain_with_grain_smaller_than_offset
```
Comment (courtneyholcomb) [inline tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_to_grain_with_grain_smaller_than_offset__query_output.txt:1]: looks right! But need where constraint to work

File: tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_window_with_custom_grain__query_output.txt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+test_name: test_offset_window_with_custom_grain
```
[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_window_with_custom_grain__query_output.txt:1]: or just add custom grain in the multiple grains query?

File: tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_window_with_date_part_only__query_output.txt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+test_name: test_offset_window_with_date_part_only
```
Comment (courtneyholcomb) [inline tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_window_with_date_part_only__query_output.txt:1]: also doesn't really make sense... remove? wait, didn't this fail? did I already remove it?

Comment (courtneyholcomb) [reply tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_window_with_date_part_only__query_output.txt:1]: yes but why did this work and not grain to date? check the rules

File: tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_window_with_grain_matching_offset__query_output.txt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+test_name: test_offset_window_with_grain_matching_offset
```
Comment (courtneyholcomb) [inline tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_window_with_grain_matching_offset__query_output.txt:1]: looks right, should add more test data

File: tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_window_with_grain_smaller_than_offset__query_output.txt [hunk 8]
```diff
@@ -0,0 +1,20 @@
+test_name: test_offset_window_with_grain_smaller_than_offset
```
Comment (courtneyholcomb) [inline tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_window_with_grain_smaller_than_offset__query_output.txt:1]: add more test data; this one should not look right

Comment (courtneyholcomb) [reply tests_metricflow/snapshots/test_offset_metrics.py/str/DuckDB/test_offset_window_with_grain_smaller_than_offset__query_output.txt:1]: test data should ensure there are bookings on the same days for 3 consecutive months

File: metricflow/dataflow/builder/dataflow_plan_builder.py [hunk 9]
```diff
@@ -1365,15 +1365,9 @@ def _build_input_measure_spec(
...
         smallest_queried_agg_time_grain: Optional[ExpandedTimeGranularity] = None
-
-        # If querying an offset metric, will join to time spine. Typically this join will happen before aggregation,
-        # since aggregation often obscures the grain needed for the offset. However, for offset_window, if the grain
-        # we'll aggregate to matches the offset grain, the results make more sense if the join happens post-aggregation,
-        # so we make an exception for that case (common for period over period metrics).
         before_aggregation_time_spine_join_description = None
         after_aggregation_time_spine_join_description = None
         if child_metric_offset_window is not None or child_metric_offset_to_grain is not None:
-            # Determine the base grain for the offset
```
Comment (courtneyholcomb) [inline metricflow/dataflow/builder/dataflow_plan_builder.py]: consolidate into earlier commit

**Comment by courtneyholcomb** [comment]
@plypaul  I specifically added rows that would be useful for testing in [this commit ](https://github.com/dbt-labs/metricflow/pull/1826/commits/ff88766bb2687d8dbb72fd1bb5ce3157167804b4#diff-7b579eb04a820634882f3b4b9ea21dec3099187f19a57e6af2ee5235604af480)(in the next PR up the stack), which ensured that the tests added would have thorough enough data to display the behavior we wanted to test. So these rows were added fairly intentionally, not arbitrarily (though with a random number of bookings per day, so that was arbitrary).

I haven't been thinking of this test data as something we need to read directly to verify the output is correct. It seems to me that our existing tests are secure enough that we can be confident that things like the output of the simple `bookings` metric are correct even when the underlying data changes. So then I found it useful to add more robust data so that we can compare more complex metric output to simple metric output. For example, in the commit linked above, this data was useful for testing day grain with month offset (requires a few days in a row of data in a few consecutive months), and for testing year grain with month offset (requires data in multiple years), etc etc. So I do think it's pretty crucial to have a robust amount of data here in order to illustrate more realistic scenarios.

**Comment by plypaul** [comment]
Yeah, a robust data set is needed, but one that is easy to check helps in the following case: I'm writing integration tests for a new feature. I author the test, and write what I think is the correct check query, and I get a mismatch. The generated query and the check query look correct, so I have to dig into the test data to see if there's an edge case that I'm not considering.

To clarify, getting better coverage takes priority, so if that's the test data that's needed to fully exercise the feature, then LGTM.

**Review by WilliamDee — APPROVED** [review]
